tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/service/impl/ProgrammeMembershipServiceImpl.java

```diff
@@ -350,4 +389,39 @@ private void updatePersonWhenStatusIsStale(Long personId) {
       personRepository.save(person);
     }
   }
+
+  /**
+   * Set a ProgrammeMembership AmendedDate from its database record.
+   *
+   * <p>The amendedDate of the programmeMembershipDto which is being saved is based on the entry
+   * in the CurriculumMembership table, which will have a slightly different AmendedDate value
+   * from the record in the ProgrammeMembership table. Since this field is used for @Version
+   * optimistic locking, we need the actual value from the ProgrammeMembership table otherwise we
+   * will get 'You are acting on stale data, please refresh' concurrency failures.
+   *
+   * <p>Caution: by overwriting the AmendedDate, it is possible to handle concurrent updates to the
+   * same ProgrammeMembership incorrectly: for exampel, if record A is updated to A' and this is
+   * committed, and then update B (which was based on A not A') is committed, this should fail as
+   * B needs to be based on A', otherwise the A' update is lost. To avoid this, apply the
+   * ProgrammeMembership record Save after the parallel CurriculumMembership Save in the same
+   * code block: in the event of stale data (the B update in the example above) the
+   * CurriculumMembership Save will fail and throw an exception before the incorrect
+   * ProgrammeMembership Save is applied.
+   *
+   * @param pm the ProgrammeMembership to set the amended date for
+   * @return the updated ProgrammeMembership
+   */
+  private ProgrammeMembership setProgrammeMembershipAmendedDateFromDbRecord(
```

Comment:
I found the UI always uses the lastModifiedDate which is received from the backend for `amendedDate` in `ProgrammeMembershipDTO`, so I'm a little wondering if the `amendedDate` needs to be set again here.

Reply:
If I understand the comment correctly, you're wondering if we need to bother setting the amendedDate, since it should be correctly passed from the UI?
The issue is that when we save() a record, it is saved to both the ProgrammeMembership and CurriculumMembership tables - but the amendedDate field in the database for each record is populated with a timestamp, which differs between the two saved records by some milliseconds. Since we then only load one version of the record (from CurriculumMembership) we get an amendedDate value from that record - and then when we later try to save an updated version of it as a ProgrammeMembership record (to keep them updated in parallel), the database rejects it because the amendedDate does not match the value it has for that record in the ProgrammeMembership table. This is part of the optimistic-locking check to make sure you are not saving an outdated version of a record over a newer version. The code here is a little bit hacky and I'm not 100% happy about it, but I can't think of an elegant alternative: loading and doing all the processing to both versions of the record (from the CurriculumMembership and ProgrammeMembership tables) seemed likely to be clumsy and verbose, but I am very open to alternatives to the method I've implemented, which I would certainly reject myself if the parallel-save code was intended to be more than temporary.